### ISSUE TITLE
prefer-spread doesn't report Reflect.apply calls.

### ISSUE THREAD

**Issue by remcohaszing** [eslint/eslint#7401]
**Tell us about your environment**
- **ESLint Version:** 3.8.1
- **Node Version:** v6.9.0
- **npm Version:** 3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

``` yaml
rules:
  prefer-spread: 2
```

**What did you do?**

From the example:

``` js
foo.apply(null, args);
```

One of my coworkers has decided to refactor this sort of syntax using `Reflect` instead of spread params.

``` js
Reflect.apply(foo, null, args);
```

**What did you expect to happen?**

I would expect `prefer-spread` to not accept this syntax either, enforcing my coworkers to use the spread syntax. At least I think it should be possible to enforce this using a setting.

**What actually happened? Please include the actual, raw output from ESLint.**

Eslint didn't report anything.

**Comment by kaicataldo** [comment]
Just a quick note - there's [discussion about deprecating this rule entirely](https://github.com/eslint/eslint/issues/7226), so this may stay as being "evaluated" until that decision is made.

**Comment by not-an-aardvark** [comment]
@kaicataldo I think that's `prefer-reflect`; this is `prefer-spread`, a different rule.

**Comment by kaicataldo** [comment]
Yikes, thanks.

**Comment by nzakas** [comment]
I'm not sure if this is a good addition to the rule. It seems like your main goal here is to disallow `Reflect.apply()`. I think you can probably accomplish that with [no-restricted-properties](http://eslint.org/docs/rules/no-restricted-properties) instead.

**Comment by nzakas** [comment]
FWIW, this is not a bug because the rule was never intended to cover this case.

**Comment by ljharb** [comment]
Was that an intentional omission though, or an unintentional one? It seems more likely that `Reflect.apply` just wasn't considered.

**Comment by nzakas** [comment]
You're correct, it wasn't considered. Considering it now, though, I don't think it makes sense to add to this rule. The goal of the rule was to point out where you could replace `.apply()` with a simple function call using rest params. I don't think `Reflect.apply()` is the same case, as I'm assuming people will mostly use that inside of proxies or will have other reasons for using it. It seems unlikely that someone would use `Reflect.apply(someFunc, null, args)` instead of `someFunc.apply(null, args)` accidentally.

**Comment by ljharb** [comment]
Fair point - I agree it's unlikely - but the OP does show that it does happen :-) I suspect a codebase where `prefer-reflect` is enabled is much more likely to overuse `Reflect` methods.

Since this can be completely covered with a `no-restricted-properties` configuration, that does seem like a sufficient workaround for the OP. However, I also don't think that it would be problematic to add to this rule - when it doesn't happen (the majority) it won't impact anyone, and when it does, it could fulfill the spirit of "prefer spread whenever you can use it" ¯\\\_(ツ)_/¯

**Comment by nzakas** [comment]
Our standard operating procedure is to keep rules as small as possible. The argument that "it won't hurt to have it in the rule" only leads to large rules that are hard to maintain. Since in this case it looks like there is another way to accomplish the same thing, it seems like not updating the rule for this specific case is the best case forward.

**Comment by kaicataldo** [comment]
Closing this issue as it looks like a sufficient workaround was suggested and it doesn't look like there's enough support to pursue an enhancement of the rule.
